### PR TITLE
feat: editable ide tabs

### LINKED ISSUE #32440: [Task] - IDE tabs revamp - editable tab name
1. Each tab name needs to be editable inside tab itself. This can be triggered from context menu as well as a single click on active tab.
2. Add unit test case to ensure the renaming functionality.

Design - https://zpl.io/AOjRmme

### PR DESCRIPTION
**Pull request by alex-golovanov**
## Description
Adds editable tab behavior for queries and JS objects.

Fixes #32440

/ok-to-test tags="@tag.All"

### :mag: Cypress test results

> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/11288430138>
> Commit: dc89acbd51afc6b238283836c6305ab68337575d
> <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=11288430138&attempt=3" target="_blank">Cypress dashboard</a>.
> Tags: `@tag.All`
> Spec:
> <hr>Fri, 11 Oct 2024 09:55:27 UTC

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [ ] No

## Release Notes

- **New Features**
	- Introduced a new dependency for improved React hooks functionality.
	- Added new constants for data source operations and user interactions, enhancing feedback and clarity.

- **Improvements**
	- Enhanced messaging related to data source actions, providing clearer user prompts and error handling messages.
	- Renamed constants for better readability and consistency across the application.

- **Bug Fixes**
	- Corrected naming conventions for constants to improve consistency across the application.

- **Chores**
	- Removed deprecated `FileTabs` component and related tests to streamline the codebase.
	- Added unit tests for the `FileTab` component to ensure expected functionality.
	- Updated the `Text` component to improve ref handling and styling options.

### CODE REVIEW

File: app/client/src/IDE/Components/FileTab/FileTab.tsx [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+  } = useBoolean(false);
+
+  const previousTitle = usePrevious(title);
+  const [editableTitle, setEditableTitle] = useState(title);
+  const currentTitle =
+    isEditing || isLoading || title !== editableTitle ? editableTitle : title;
+  const [validationError, setValidationError] = useState<string | null>(null);
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  const attemptToSave = () => {
```
Comment (sagar-qa007) [inline app/client/src/IDE/Components/FileTab/FileTab.tsx:52]: Did we considering adding unit test cases for this PR?

**Review by sagar-qa007 — CHANGES_REQUESTED** [review]

File: app/client/src/IDE/Components/FileTab/FileTab.tsx (hunk 1 above)
Comment (alex-golovanov) [reply app/client/src/IDE/Components/FileTab/FileTab.tsx:52]: Yes, there will be unit tests for this component.

**Review by KelvinOm — DISMISSED** [review]

**Review by hetunandu — DISMISSED** [review]

File: app/client/src/selectors/ui.tsx [hunk 2]
```diff
@@ -34,6 +38,10 @@ export const getErrorForApiName = (state: AppState, id: string) =>
 export const getIsSavingForJSObjectName = (state: AppState, id: string) =>
   state.ui.jsObjectName.isSaving[id];
 
+/** Select saving status for all JS object names */
+export const getJsObjectNameSavingStatuses = (state: AppState) =>
```
Comment (ankitakinger) [inline app/client/src/selectors/ui.tsx:42]: Just out of curiosity, why is the name ending in plural "statuses"? Shouldn't it be just "status"?

Comment (alex-golovanov) [reply app/client/src/selectors/ui.tsx:42]: Well, there are plural statuses. If this is about plural form of the word `status`, then I feel that in this case explicit `statuses` is more fitting/descriptive and conveys the purpose of the variable without further need to investigate whether it's single or plural. Would you agree?

Comment (ankitakinger) [reply app/client/src/selectors/ui.tsx:42]: Alright. Makes sense. I just checked the code to understand why this state is plural.

**Review by ankitakinger — DISMISSED** [review]

**Review by albinAppsmith — APPROVED** [review]

**Comment by albinAppsmith** [comment]
@alex-golovanov @momcilo-appsmith check this behaviour. Maybe we should cancel name edit if field is not valid and save it if field is valid.

https://github.com/user-attachments/assets/e37279d7-9f1f-4bf1-a22f-dda67e1b9502

**Review by hetunandu — APPROVED** [review]

**Review by albinAppsmith — APPROVED** [review]

**Review by ankitakinger — APPROVED** [review]

**Review by sagar-qa007 — APPROVED** [review]

**Pull request merged**